### PR TITLE
Add DataOutResample

### PR DESCRIPTION
**Pull request by peterrum**
FYI @mschreter

### CODE REVIEW

File: include/deal.II/numerics/data_out_resample.h
Comment (bangerth) [inline include/deal.II/numerics/data_out_resample.h:35]:
```suggestion
 * one can output the result obtained on an unstructured mesh onto a structured one
```

File: include/deal.II/numerics/data_out_resample.h
Comment (bangerth) [inline include/deal.II/numerics/data_out_resample.h:65]: I think it might be nice if this function took the same arguments as `DataOut::build_patches()`, and then just forward these arguments.

File: include/deal.II/numerics/data_out_resample.h
Comment (bangerth) [inline include/deal.II/numerics/data_out_resample.h:75]: make this a `SmartPointer`.

File: include/deal.II/numerics/data_out_resample.h
Comment (bangerth) [inline include/deal.II/numerics/data_out_resample.h:85]: Same here

File: include/deal.II/numerics/data_out_resample.h
Comment (bangerth) [inline include/deal.II/numerics/data_out_resample.h:100]: and this should probably be a unique_ptr

File: source/numerics/data_out_resample.cc
Comment (bangerth) [inline source/numerics/data_out_resample.cc:55]: This would be easier to understand if you just used
```
  const Quadrature<patch_dim> quadrature (fe.unit_support_points());
```
which I think is the same.

File: source/numerics/data_out_resample.cc
Comment (bangerth) [inline source/numerics/data_out_resample.cc:72]: make this a unique_ptr

File: source/numerics/data_out_resample.cc
Comment (bangerth) [inline source/numerics/data_out_resample.cc:91]: don't you also need to consider the case that `a.first == b.first`?

File: source/numerics/data_out_resample.cc
Comment (bangerth) [inline source/numerics/data_out_resample.cc:130]: maybe give this variable a more descriptive name?

File: source/numerics/data_out_resample.cc
Comment (bangerth) [inline source/numerics/data_out_resample.cc:162]: this would be the place to forward whatever arguments `build_patches()` got.

File: include/deal.II/numerics/data_out_resample.h
Comment (bangerth) [inline include/deal.II/numerics/data_out_resample.h:105]: can you give this variable a more descriptive name?

**Review by bangerth — COMMENTED** [review]
I *think* that you broke the interface into `build_patches()` and `update_mapping()` because you want to re-use objects across multiple calls to `build_patches()`, but I don't actually know. My preference would be to drive this class in the same way as `DataOut`, where one can/should not re-use the object after generating graphical output once. In that case, `update_mapping()` would become an internal function called only once in `build_patches()`, or would in fact be wholly subsumed by the `build_patches()` function. In any case, I don't actually think I know what the function does or is supposed to do because it hasn't become clear to me from its name *which* *mapping* is being updated or should be updated. Maybe a different name would help?

Alternatively, one could imagine passing a `Utilities::MPI::RemotePointEvaluation` as argument that the function modifies as necessary and that can carry state to the next invocation, if anyone wants that. This would obviate the need to keep the `DataOutResample` object around.

File: tests/remote_point_evaluation/data_out_resample_01.cc
Comment (peterrum) [reply tests/remote_point_evaluation/data_out_resample_01.cc]: The vector needs a `std::shared_ptr`.

File: source/numerics/data_out_resample.cc
Comment (peterrum) [reply source/numerics/data_out_resample.cc:91]: Why? std::sort expects a lambda for `<`. In the next step, when I remove duplicates I am checking `==`.

File: source/numerics/data_out_resample.cc
Comment (peterrum) [reply source/numerics/data_out_resample.cc:55]: True :+1:

**Comment by peterrum** [comment]
@bangerth Thanks for your review! I have incorporated most points. Personally, I would like to keep `update_mapping()` (motivated by `MatrixFree::update_mapping()`) and `build_patches()`. Doing the first is not cheap (since the processes/cells have to be identified that own the points) and should be done as few time as possible. 

Given the responsibility of RPE to the user in not really an option because `partitioner` and `point_to_local_vector_indices` are also needed.

File: source/numerics/data_out_resample.cc
Comment (bangerth) [reply source/numerics/data_out_resample.cc:84]: I was just imagining that you are doing a lexicographic ordering. But maybe you don't :-)

File: include/deal.II/numerics/data_out_resample.h
Comment (kronbichler) [inline include/deal.II/numerics/data_out_resample.h:33]:
```suggestion
 * A DataOut class that does not output a numerical solution on the cells of the original triangulation
```

File: include/deal.II/numerics/data_out_resample.h
Comment (peterrum) [reply include/deal.II/numerics/data_out_resample.h:68]: I can do this.

File: source/numerics/data_out_resample.cc
Comment (peterrum) [reply source/numerics/data_out_resample.cc:49]: Shouldn't FE_Q not be more exact? Later on, the actual DataOut object will do an interpolation itself.

File: source/numerics/data_out_resample.cc
Comment (peterrum) [reply source/numerics/data_out_resample.cc:113]: Mapping is a smart pointer. So it will complain. But I guess I can add an additional assert here.

File: source/numerics/data_out_resample.cc
Comment (peterrum) [reply source/numerics/data_out_resample.cc:135]: I was thinking about the same. But I am not sure how to accomplish this yet. There are two steps right now in:

https://github.com/dealii/dealii/blob/74061654d721022260ea6eb633942e98059468b9/include/deal.II/numerics/vector_tools_evaluate.h#L299-L306

1) read dof-values by FEValues and 2) evaluate FEPointEvaluation. Would I add a step 1a) that picks all dofs of the first component if a component is given by the user that is not -1. But I would like to keep this out of this PR.

File: source/numerics/data_out_resample.cc
Comment (peterrum) [reply source/numerics/data_out_resample.cc:98]: Not in this PR ;) but yes we should do that - it is not the first time that I have implemented these lines of codes and there are copies in too many projects!

**Comment by peterrum** [comment]
@kronbichler Thanks for the review. I will try to integrate the changes before the workshop!

File: source/numerics/data_out_resample.cc
Comment (kronbichler) [reply source/numerics/data_out_resample.cc:49]: As far as I can see, you later evaluate the patches on exactly those points, so my assumption would have been that this one does no additional interpolation besides the place where we go from the original triangulation to the helper DoFHandler, whereas the other does for more than 2 patches.

Comment (peterrum) [reply source/numerics/data_out_resample.cc:49]: I guess you are right. This is what `QIteratated` does more or less.

**Comment by peterrum** [comment]
@kronbichler I have made the changes!

**Review by tjhei — COMMENTED** [review]
I forgot to submit my review from a couple of days ago. Oops.

File: include/deal.II/numerics/data_out_resample.h
Comment (peterrum) [reply include/deal.II/numerics/data_out_resample.h:53]: By now I have added two notes.

File: include/deal.II/numerics/data_out_resample.h
Comment (peterrum) [reply include/deal.II/numerics/data_out_resample.h:60]: Done.

File: include/deal.II/numerics/data_out_resample.h
Comment (peterrum) [reply include/deal.II/numerics/data_out_resample.h]: Done.

File: include/deal.II/numerics/data_out_resample.h
Comment (peterrum) [reply include/deal.II/numerics/data_out_resample.h:38]: Done.

File: include/deal.II/numerics/data_out_resample.h
Comment (peterrum) [reply include/deal.II/numerics/data_out_resample.h:48]: Patch is the term in the library: this is the triangulation on which the patches are created on. I have updated the comment.

**Comment by peterrum** [comment]
The failing test is unrelated!

File: include/deal.II/numerics/data_out_resample.h
Comment (bangerth) [inline include/deal.II/numerics/data_out_resample.h:33]:
```suggestion
 * A DataOut-like class that does not output a numerical solution on
```

File: include/deal.II/numerics/data_out_resample.h
Comment (bangerth) [inline include/deal.II/numerics/data_out_resample.h:74]:
```suggestion
   * Just like in the DataOut::build_patches() context,  @p n_subdivisions determines how many
```

File: include/deal.II/numerics/data_out_resample.h
Comment (bangerth) [inline include/deal.II/numerics/data_out_resample.h:90]: This would also be useful to state in the documentation of `build_patches()`: That the function re-computes where all of the evaluation nodes are.

File: include/deal.II/numerics/data_out_resample.h
Comment (bangerth) [inline include/deal.II/numerics/data_out_resample.h:93]: Maybe also mention that it uses the `n_subdivisions` previously passed to these functions.

File: source/numerics/data_out_resample.cc
Comment (bangerth) [reply source/numerics/data_out_resample.cc:130]: I think this would be good.

File: source/numerics/data_out_resample.cc
Comment (bangerth) [inline source/numerics/data_out_resample.cc:129]: Does rpe attach to signals that invalidate data whenever the triangulation changes?

File: source/numerics/data_out_resample.cc
Comment (bangerth) [inline source/numerics/data_out_resample.cc:169]: Why not let the user provide the names (and, for that matter, component interpretation) of vectors? It seems unfriendly to just call all output variables `temp_*`.

Separately, this is a really difficult to find place where the counter is incremented. You're very much asking for mistakes if someone uses the counter below if the code below is ever changed. Please move the increment of `counter` to the very last line of the loop.

**Review by bangerth — COMMENTED** [review]
Close. I'm still hung up on `update_mapping()`, though. I *think* that there is an implicit assumption somewhere that in typical uses, the class itself sticks around and will be re-used -- contrary to how one generally does this for `DataOut`. One would then before calling `build_patches()` also call `update_mapping()` (or would one only do that the second and all following times?).

The problem I'm having, not knowing the typical workflow for matrixfree stuff, is that none of these implicit assumptions are documented, and as a consequence I cannot figure out in which contexts one would call `update_mapping()` and in which contexts that can be avoided. 

I think it would be useful if you added a small code example or other explanation to the class declaration that discusses the use cases you see, and how that would be translated into function calls.

File: source/numerics/data_out_resample.cc
Comment (peterrum) [reply source/numerics/data_out_resample.cc:129]: Yes.

File: source/numerics/data_out_resample.cc
Comment (peterrum) [reply source/numerics/data_out_resample.cc:169]: > Why not let the user provide the names (and, for that matter, component interpretation) of vectors? It seems unfriendly to just call all output variables temp_*.

The variable names are not relevant here. It is actually not used, since the actual output of the data is not done by `patch_data_out`.

Comment (bangerth) [reply source/numerics/data_out_resample.cc:169]: Ah, I see. In that case, please add a comment to that effect.

**Comment by peterrum** [comment]
@bangerth I have made the changes!

File: source/numerics/data_out_resample.cc
Comment (kronbichler) [reply source/numerics/data_out_resample.cc:157]: OK, we can extend this at a later time with more experience on this topic.

**Review by kronbichler — COMMENTED** [review]
I am in support of this, but I am biased here, so we need another approving reviewer.

**Comment by peterrum** [comment]
What about merging this? I have 1-2 follow-up PRs planned.

**Review by kronbichler — APPROVED** [review]
I do approve, but let's wait for another review.

**Comment by kronbichler** [comment]
@tjhei @bangerth any final comments? If not, I suggest we merge tomorrow.

**Pull request merged**
